TE-14.2 fix logics & added repaired_vrf

snaragund opened the pull request:
- Added vrf_repaired as vrf_r config without vrf_rd is not logical.
- Under deviations.ExplicitInterfaceInDefaultVRF(dut) included DefaultVRFIPv4NHCount while configuring NHs.
-  Redefined new deviations.DecapNHWithoutNextHopNIUnsupported, refer PR: https://github.com/openconfig/featureprofiles/pull/5637. 
- Under BuildVRFCfonfig function added missing code which was included in TE-14.1.
- For Nokia vendor added decapIPv4ScaleCount=128. & using `override_default_nh_scale` deviation for same: https://partnerissuetracker.corp.google.com/u/0/issues/443020838#comment17

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
TE-14.2 fix logics & added repaired_vrf
- Added vrf_repaired as vrf_r config without vrf_rd is not logical.
- Removed ! with deviations.DecapNHWithNextHopNIUnsupported.
- Under deviations.ExplicitInterfaceInDefaultVRF(dut) included DefaultVRFIPv4NHCount while configuring NHs.
- Under BuildVRFCfonfig function added missing code which was included in TE-14.1.
- For Nokia vendor added decapIPv4ScaleCount=128.
- Added `override_default_nh_scale` deviation.
- Existing scale.go configures 29254 NHs where as many NHs are unused so changed logic to configure NHs which are needed(9511 NHs).

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
Merge branch 'main' into SND-te14dot2

snaragund committed:
Merge branch 'main' into SND-te14dot2

ram-mac commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -597,7 +598,7 @@ func pushDecapScaleEntries(t *testing.T, args *testArgs, decapEntries []string)
 func installDecapEntry(t *testing.T, args *testArgs, nhIndex, nhgIndex uint64, prefix string) {
 	decapNH := fluent.NextHopEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(args.dut)).
 		WithIndex(nhIndex).WithDecapsulateHeader(fluent.IPinIP)
-	if !deviations.DecapNHWithNextHopNIUnsupported(args.dut) {
```
@snaragund - Why are we removing this deviation statement? Also, I see this deviation is not used by Nokia. This  break other vendor tests. Can we please revert this change?

AmrNJ replied:
@snaragund This breaks the test for vendors where this deviation is enabled. The logic should remain unchanged if Nokia supports it.

snaragund committed:
-Reverting back NH & NHGs optimization as per comment by Ram.
-Reverting changes of deviation definition as per comment by Ram.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
Merge branch 'main' into SND-te14dot2

snaragund committed:
Merge branch 'main' into SND-te14dot2

snaragund committed:
-Reverted change without !

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
Merge branch 'main' into SND-te14dot2

snaragund committed:
Merge branch 'main' into SND-te14dot2

snaragund committed:
Merge branch 'main' into SND-te14dot2

AmrNJ commented:
@snaragund Please can you address the comments?

snaragund committed:
1. Added if else statement for DecapNHWithNextHopNIUnsupported instead of !.
2. Added deviation in Nokia vendor.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
Merge branch 'main' into SND-te14dot2

AmrNJ commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -597,7 +598,9 @@ func pushDecapScaleEntries(t *testing.T, args *testArgs, decapEntries []string)
-	if !deviations.DecapNHWithNextHopNIUnsupported(args.dut) {
+	if deviations.DecapNHWithNextHopNIUnsupported(args.dut) {
+		// Omit WithNextHopNetworkInstance (metadata decap_nh_with_nexthop_ni_unsupported).
+	} else {
```
Both the logic are functionally same. Its better to have !deviations as it is more cleaner.

snaragund replied:
Based on deviations guidelines removed ! usage & replaced it with if/else condition.

AmrNJ commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
Add a 15s sleep here for the counters to be updated.

snaragund replied:
In Nokia platform 15sec sleep it was fine, if this change is needed for other vendors then will make it to 30sec.

AmrNJ replied:
15s is fine and is already tested across all vendors.

AmrNJ commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
This should be t.Fatalf

snaragund replied:
Error is ok test finally fails after both tests completed.

AmrNJ replied:
The logs would be misleading if this is not Fatalf. Though the txPackets would be 0, the log would say that there is no traffic loss. Please change it.

AmrNJ replied:
Resolving it as this is not related to the PR change.

AmrNJ commented on internal/tescale/scale.go:
```diff
@@ -226,9 +226,9 @@ func BuildVRFConfig(dut *ondatra.DUTDevice, egressIPs []string, param Param) []*
 			vrfDefault.NHGs = append(vrfDefault.NHGs, nhgEntry)
 			nhgID = idPool.NextNHGID()
-			nhgEntry = fluent.NextHopGroupEntry().WithID(nhgID).WithNetworkInstance(defaultVRF).WithBackupNHG(nhgDecapToDefault)
+			nhgEntry = fluent.NextHopGroupEntry().WithID(nhgID).WithNetworkInstance(defaultVRF).WithBackupNHG(nhgDecapToDefault).AddNextHop(nhID, 1)
```
This change is not needed. The code passes regardless.

snaragund replied:
Optimization was for boundary & non-boundary scenarios. Reverted.

snaragund committed:
- Made changes as per reviewers suggestion to increase sleep to 30sec & removed optimization for NHG/NH creation

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

snaragund committed:
-Reverted back to 15sec as change was not required.

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

gggsmith commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -822,16 +825,19 @@ func configureDUTSubIfs(t *testing.T, dut *ondatra.DUTDevice, dutPort *ondatra.P
+			index := uint32(i)
+			fptest.AssignToNetworkInstance(t, dut, dutPort.Name(), deviations.DefaultNetworkInstance(dut), index)
+		}
+	}
```
For my understanding, could you please clarify - why the sub-interface assignment has to be moved out of the loop?

snaragund replied:
Assignment used to run inside the main loop before batchConfig.Set, so VRF attachment could happen before batched L3 + neighbor config was committed. Above changes ensures subinterfaces (and ARP) exist on the DUT before subinterfaces are bound to DEFAULT, which matches common implementation expectations and avoids ordering-sensitive failures.

gggsmith replied:
Thank you for the clarification!

gggsmith commented on feature/gribi/otg_tests/encap_decap_scale/metadata.textproto:
```diff
@@ -25,6 +25,8 @@ platform_exceptions:  {
     explicit_interface_in_default_vrf:  true
     interface_enabled:  true
+    override_default_nh_scale: true
+    decap_nh_with_nexthop_ni_unsupported: true
```
Could you please point me out to a reference to the relevant FR?

snaragund replied:
In discussion with PLM to raise buganizer.

AmrNJ replied:
@snaragund Please can you tag the issue tracker here?

snaragund replied:
@AmrNJ after merging of https://github.com/openconfig/featureprofiles/pull/5637 will push respective changes here.

snaragund replied:
https://partnerissuetracker.corp.google.com/u/0/issues/443020838#comment17 tracker for using `override_default_nh_scale` deviation.

gggsmith replied:
For posterity - that is the 393162725 issue. Thank you!

gggsmith commented on feature/gribi/otg_tests/encap_decap_scale/encap_decap_scale_test.go:
```diff
@@ -1063,5 +1084,8 @@ func overrideScaleParams(dut *ondatra.DUTDevice) {
 			decapIPv4ScaleCount = 400
 		}
+		if dut.Vendor() == ondatra.NOKIA {
+			decapIPv4ScaleCount = 128
```
To confirm - rn Nokia supports 48 decap ipv4 entries 48 of mixed (/22 /24 /26 /28) plus 128 /32 on top of it ?

snaragund replied:
Nokia supports max of 128 decap prefixes irrespective of subnet length. Moreover as you can see in test with override_default_nh_scale deviation for Nokia, it programs max of 128 because we do flush before programming scaled decap entries.

gggsmith reviewed (changes requested):
Thank you for working on this Sham!

One more question I had - I see you mentioned in the description:
> Existing scale.go configures 29254 NHs where as many NHs are unused so changed logic to configure NHs which are needed(9511 NHs).

And i don't see that change in the PR. Is that something I missed or is that became obsolete?

snaragund commented:
> Thank you for working on this Sham!
> 
> One more question I had - I see you mentioned in the description:
> 
> > Existing scale.go configures 29254 NHs where as many NHs are unused so changed logic to configure NHs which are needed(9511 NHs).
> 
> And i don't see that change in the PR. Is that something I missed or is that became obsolete?

Yes that optimization code was removed as suggested by @ram-mac

AmrNJ commented:
Hi @snaragund Please can you help populate the issue tracker IDs?

snaragund commented:
@gggsmith @AmrNJ any updates? If #5637 is merged on priority then will incorporate changes in this accordingly.

snaragund committed:
Merge branch 'main' into SND-te14dot2

alokmtri-g commented:
/fptest nokia-physical

alokmtri-g commented:
Test Status: Passed 
Test Run log: https://fusion2.corp.google.com/invocations/4b8ff086-4a93-4d00-8abb-e9a78513ef12/targets/%2F%2Fthird_party%2Fopenconfig%2Ffeatureprofiles%2Ffeature%2Fgribi%2Fotg_tests%2Fencap_decap_scale:nokia_7250_ixr10e/tests

navaneethyv reviewed (approved):
Approving as per discussion with alok.
